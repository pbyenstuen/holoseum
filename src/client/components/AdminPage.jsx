import React from "react";
import UploadForm from "./UploadForm";
import HologramList from "./HologramList";
import useLoader from "../hooks/useLoader";

const AdminPage = ({ api }) => {
    const { data: holograms, loading, error, reload: updateList } = useLoader(async () => await api.holo.getHolograms());

    return (
        <div id="admin-page-container">
            <UploadForm
                api={api}
                updateList={updateList} />
            <HologramList
                api={api}
                holograms={holograms}
                loading={loading}
                error={error}
                updateList={updateList}
            />
        </div>
    );
};

export default AdminPage;