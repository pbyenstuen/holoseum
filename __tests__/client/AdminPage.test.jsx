import React from "react";
import ReactDOM from "react-dom";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import AdminPage from "../../src/client/components/AdminPage";

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
    });
}

describe("AdminPage", () => {
    it("shows uploaded files", async () => {
        const api = {
            holo: {
                getHolograms: () => [{
                    _id: 0,
                    metadata: "Skipskranen"
                }]
            }
        }

        await mountComponent(<AdminPage api={api} />);

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("p").textContent).toEqual("Skipskranen");
    });

    it("can upload new file", async () => {
        const uploadHologram = jest.fn();
        const file = new File(["tollvaktboden"], "tollvaktboden.mp4", { type: "video/mp4" });

        const api = {
            holo: {
                uploadHologram,
                getHolograms: () => [{
                    _id: 0,
                    metadata: "Skipskranen"
                }]
            }
        }

        await mountComponent(<AdminPage api={api} />);

        Simulate.change(container.querySelector("input[name='name']"), {
            target: { value: "Tollvaktboden" },
        });

        Simulate.change(container.querySelector("input[name='file']"), {
            target: { files: [file] },
        });

        await act(async () => {
            await Simulate.submit(container.querySelector("form"));
        });

        expect(uploadHologram).toBeCalled();
    });

    it("can show error message on uploading file", async () => {
        const file = new File(["tollvaktboden"], "tollvaktboden.mp4", { type: "video/mp4" });

        const api = {
            holo: {
                getHolograms: () => [{
                    _id: 0,
                    metadata: "Skipskranen"
                }],
                uploadHologram: () => {
                    throw new Error("Error");
                }
            }
        }

        await mountComponent(<AdminPage api={api} />);

        Simulate.change(container.querySelector("input[name='name']"), {
            target: { value: "Tollvaktboden" },
        });

        Simulate.change(container.querySelector("input[name='file']"), {
            target: { files: [file] },
        });

        await act(async () => {
            await Simulate.submit(container.querySelector("form"));
        });

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h4").textContent).toEqual("Error: Error");
    });

    it("can delete file", async () => {
        const deleteHologram = jest.fn();

        const api = {
            holo: {
                deleteHologram,
                getHolograms: () => [{
                    _id: 0,
                    metadata: "Skipskranen"
                }]
            }
        }

        await mountComponent(<AdminPage api={api} />);

        await act(async () => {
            await Simulate.click(container.querySelector("#del-btn"));
        });

        expect(deleteHologram).toBeCalledWith({
            name: "Skipskranen"
        });
    });

    it("can show error message for empty fields", async () => {
        await mountComponent(<AdminPage />);

        Simulate.submit(container.querySelector("form"));

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h4").textContent).toEqual("Fyll inn alle feltene");
    });
});