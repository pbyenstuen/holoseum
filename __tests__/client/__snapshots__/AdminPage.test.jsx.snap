// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminPage can show error message for empty fields 1`] = `"<div id=\\"admin-page-container\\"><section id=\\"upload-form-container\\"><h2>Last opp hologram</h2><div id=\\"upload-form-card\\" class=\\"card\\"><form><h4>Fyll inn alle feltene</h4><div><div><label>Navn <input name=\\"name\\" type=\\"text\\" value=\\"\\"></label></div><div><label>Hologram <input name=\\"file\\" type=\\"file\\" value=\\"\\"></label></div><button type=\\"submit\\">LAST OPP</button></div></form></div></section><section id=\\"hologram-list-container\\"><h2>Opplastinger</h2><div><p><i>Ingenting Ã¥ vise</i></p></div></section></div>"`;

exports[`AdminPage can show error message on uploading file 1`] = `"<div id=\\"admin-page-container\\"><section id=\\"upload-form-container\\"><h2>Last opp hologram</h2><div id=\\"upload-form-card\\" class=\\"card\\"><form><h4>Error: Error</h4><div><div><label>Navn <input name=\\"name\\" type=\\"text\\" value=\\"Tollvaktboden\\"></label></div><div><label>Hologram <input name=\\"file\\" type=\\"file\\" value=\\"\\"></label></div><button type=\\"submit\\">LAST OPP</button></div></form></div></section><section id=\\"hologram-list-container\\"><h2>Opplastinger</h2><div><article><p>Skipskranen</p><button id=\\"del-btn\\">SLETT</button></article></div></section></div>"`;

exports[`AdminPage shows uploaded files 1`] = `"<div id=\\"admin-page-container\\"><section id=\\"upload-form-container\\"><h2>Last opp hologram</h2><div id=\\"upload-form-card\\" class=\\"card\\"><form><div><div><label>Navn <input name=\\"name\\" type=\\"text\\" value=\\"\\"></label></div><div><label>Hologram <input name=\\"file\\" type=\\"file\\" value=\\"\\"></label></div><button type=\\"submit\\">LAST OPP</button></div></form></div></section><section id=\\"hologram-list-container\\"><h2>Opplastinger</h2><div><article><p>Skipskranen</p><button id=\\"del-btn\\">SLETT</button></article></div></section></div>"`;
